pkg/kubectl/resource_columns.go

```diff
@@ -0,0 +1,551 @@
+
+		if content.Kind() != reflect.Func {
+			err = fmt.Errorf("invalid print handler. %#v is not a function", entry.content)
+			return
+		}
+		funcType := content.Type()
+
+		args := []reflect.Value{}
+		for i := 0; i < funcType.NumIn(); i++ {
+			if funcType.In(i) == objType {
+				args = append(args, vof(obj))
+				continue
+			}
+			if funcType.In(i) == reflect.TypeOf((*runtime.Object)(nil)).Elem() {
+				args = append(args, vof(obj))
+				continue
+			}
+			if funcType.In(i) == contextType {
+				args = append(args, vof(c))
+				continue
+			}
+			err = fmt.Errorf("Cannot populate type %v for %s", funcType.In(i), entry.label)
+			return
+		}
+		resultValue := content.Call(args)
+
+		if funcType.Out(0) == reflect.TypeOf((*string)(nil)).Elem() {
+			result = append(result, []string{resultValue[0].String()})
+		} else if funcType.Out(0) == reflect.TypeOf((*[]string)(nil)).Elem() {
+			allRows := []string{}
+			for i := 0; i < resultValue[0].Len(); i++ {
+				allRows = append(allRows, resultValue[0].Index(i).String())
+			}
+			result = append(result, allRows)
+		} else if funcType.Out(0) == reflect.TypeOf((*error)(nil)).Elem() {
+			if !resultValue[0].IsNil() {
+				err, _ = resultValue[0].Interface().(error)
+				return
+			}
+		} else {
+			err = fmt.Errorf("Wrong return type %v for %s", funcType.In(0), entry.label)
+			return
+		}
+
+		if funcType.NumOut() == 2 {
+			if funcType.Out(1) == reflect.TypeOf((*error)(nil)).Elem() {
+				if !resultValue[1].IsNil() {
+					err, _ = resultValue[1].Interface().(error)
+					return
+				}
+			} else {
+				err = fmt.Errorf("Wrong return type %v for %s", funcType.In(1), entry.label)
+				return
+			}
+		}
+	}
+	return
+}
+
+var columnMap = map[reflect.Type][]columnDefinition{
+	reflect.TypeOf((*api.Pod)(nil)):                   podColumnsFunc,
+	reflect.TypeOf((*api.PodTemplate)(nil)):           podTemplateColumnsFunc,
+	reflect.TypeOf((*api.ReplicationController)(nil)): replicationControllerColumnsFunc,
+	reflect.TypeOf((*api.Service)(nil)):               serviceColumnsFunc,
+	reflect.TypeOf((*api.Endpoints)(nil)):             endpointsColumnsFunc,
+	reflect.TypeOf((*api.Node)(nil)):                  nodeColumnsFunc,
+	reflect.TypeOf((*api.Event)(nil)):                 eventColumnsFunc,
+	reflect.TypeOf((*api.LimitRange)(nil)):            limitRangeColumnsFunc,
+	reflect.TypeOf((*api.ResourceQuota)(nil)):         resourceQuotaColumnsFunc,
+	reflect.TypeOf((*api.Namespace)(nil)):             namespaceColumnsFunc,
+	reflect.TypeOf((*api.Secret)(nil)):                secretColumnsFunc,
+	reflect.TypeOf((*api.ServiceAccount)(nil)):        serviceAccountColumnsFunc,
+	reflect.TypeOf((*api.PersistentVolume)(nil)):      persistentVolumeColumnsFunc,
+	reflect.TypeOf((*api.PersistentVolumeClaim)(nil)): persistentVolumeClaimColumnsFunc,
+	reflect.TypeOf((*api.ComponentStatus)(nil)):       componentStatus,
+}
+
+func getColumnsFor(obj runtime.Object, c printContext) (columns []columnDefinition, err error) {
```

Comment:
@deads2k 
As above, this `ColumnPrinter` is in fact `ColumnContent` and should not be reused as the implementation is always bound to a particular instance of `runtime.Object`.

With
``` go
type ColumnPrinter interface{
    Title() string
    Content(runtime.Object) string
}
```
you will need:
```go
type FuncTColumnPrinter struct{
    title string
    content func(obj T) string
}
```
And then:
* if T is `runtime.Object` then each function will need a cast what is ugly and not particularly type safe. 
* if T is a final api object then you will need lots of boiler plate code - separate printer for each of the types.
* if content can be an arbitrary function then we will have reflection. We can ensure that all functions are coherent (accept the same type) during the initialization.

If we wanted to pass object to the content function then I would prefer to go with the last option.